### DISCUSSION TITLE
Inherit constructor

### DISCUSSION THREAD

**Discussion by PlayCreatively** [dotnet/csharplang#4641] [Q&A]
I'm trying to create a system close to Unity's with GameObjects and Components. 
`Component` is an abstract class with a constructor which requires a reference to a GameObject as a parameter. 

``` c#
public abstract class Component
{
        public readonly GameObject gameObject;

        public Component(GameObject gameObject)
        {
	        this.gameObject = gameObject;
        }
}
```

Everything is sound so far but now I've created useless boilerplate work for every single class that will ever inherit from Component. Now I have to explicitly create a constructor which takes in the arguments for the `base` class, regardless if I need a special constructor for the class.

``` c#
public class Health : Component
{
	public Health(GameObject gameObject) : base(gameObject) 
	{

	}
}
```

Why can't I have the 'inheriting' class inherit the constructor as well. Why do I have to write it out every single time. 

If there's a work around I'd greatly appreciate a comment :)
Thanks for reading.

**Comment by TonyValenti** [comment]
Unfortunately, the way that C# works today, if you require parameters to a constructor, every derived class must reimplement the constructor. 

I think that required properties (C# 10?) will solve this problem.

**Reply by PlayCreatively** [reply]
I looked it up #3630 and yes, required properties sounds like the perfect solution. Thanks for the info :)

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/806

**Comment by jnm2** [comment]
If you can use source generators in that project, this is something you could implement yourself and start using now rather than waiting many years for the language process (if the idea ever rises to priority). You could write a source generator that automatically adds constructors matching the base constructors to every class you declare. You could make it controllable using an attribute you define as internal if you prefer.

**Reply by PlayCreatively** [reply]
Yes, this does look like a typical source generator contender, but it's a beast I have yet to tame. Wish me luck 😣
